src/main/java/gg/projecteden/nexus/features/listeners/Sleep.java

```diff
@@ -16,115 +14,92 @@
-			lastCalculatedSleeping = sleeping;
-			return;
-		}
-
-		int needed = (int) Math.ceil(active / 2d);
-
-		if (sleeping != lastCalculatedSleeping || needed != lastCalculatedNeeded)
-			for (Player player : players)
-				Tasks.wait(2, () -> ActionBarUtils.sendActionBar(player, "Sleepers needed to skip night: &e" + sleeping + "&3/&e" + needed));
-
-		lastCalculatedSleeping = sleeping;
-		lastCalculatedNeeded = needed;
-
-		if (sleeping >= needed) {
-			handling = true;
-			world.setStorm(false);
-			world.setThundering(false);
-
-			players.forEach(player -> PlayerUtils.send(player, PREFIX + "The night was skipped because 50% of players slept!"));
-
-			int wait = 0;
-			while (true) {
-				long newTime = world.getTime() + (++wait * speed);
-				if (!new NumberRange(12541L, (24000L - speed)).containsNumber(newTime))
-					break;
-
-				Tasks.wait(wait, () -> world.setTime(newTime));
+	private static final long SPEED = 150;
+
+	private enum State { SLEEPING, SKIPPING }
+	private static final Map<World, State> sleepingWorlds = new HashMap<>();
+
+	static {
+		Tasks.repeatAsync(0, 1, () -> {
+			for (World world : sleepingWorlds.keySet()) {
+				long sleeping = world.getPlayers().stream().filter(player -> player.isSleeping() && canSleep(player)).count();
+				long active = world.getPlayers().stream().filter(Sleep::canSleep).count();
+				int needed = (int) Math.ceil(active / 2d);
+
+				if (sleeping >= needed && sleepingWorlds.get(world) != State.SKIPPING)
+					skipNight(world);
+				else if (sleepingWorlds.get(world) == State.SLEEPING)
+					world.getPlayers().forEach(player -> ActionBarUtils.sendActionBar(player,
+						"Sleepers needed to skip night: &e" + sleeping + "&3/&e" + needed));
 			}
-
-			Tasks.wait(wait, () -> {
-				world.setTime(0);
-				if (world.hasStorm())
-					world.setStorm(false);
-				if (world.isThundering())
-					world.setThundering(false);
-
-				handling = false;
-				lastCalculatedSleeping = 0;
-				lastCalculatedNeeded = 0;
-			});
-		}
+		});
 	}
 
-	protected boolean canSleep(Player player) {
+	private static boolean canSleep(Player player) {
 		return !PlayerUtils.isVanished(player) && !AFK.get(player).isTimeAfk() && player.getGameMode() == GameMode.SURVIVAL;
 	}
 
-	protected void handle(World world) {
-		if (!handling)
-			Tasks.wait(1, () -> calculate(world));
-	}
+	private static void skipNight(World world) {
+		sleepingWorlds.put(world, State.SKIPPING);
+		world.getPlayers().forEach(
+			player -> PlayerUtils.send(player, PREFIX + "The night was skipped because 50% of players slept!"));
 
-	protected void handle(HasPlayer player) {
-		handle(player.getPlayer().getWorld());
-	}
+		world.setStorm(false);
+		world.setThundering(false);
 
-	@EventHandler
-	public void onBedEnter(PlayerBedEnterEvent event) {
-		handle(event);
-	}
+		int emptyActionbarTaskId = Tasks.repeatAsync(0, 1, () ->
+			world.getPlayers().forEach(player -> ActionBarUtils.sendActionBar(player, " ")
+		));
+		int wait = 0;
```

Comment:
Super picky formatting thing
```suggestion
		int emptyActionbarTaskId = Tasks.repeatAsync(0, 1, () ->
			world.getPlayers().forEach(player -> ActionBarUtils.sendActionBar(player, " ")));
		
		int wait = 0;
```